### PR TITLE
FIX: applies a unique index for accepted_answer_post_id custom field

### PR DESCRIPTION
**Pull request by jjaffeux**

### CODE REVIEW

File: db/migrate/20191209095548_ensures_unique_accepted_answer_post_id.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
+
+class EnsuresUniqueAcceptedAnswerPostId < ActiveRecord::Migration[6.0]
```
Comment (davidtaylorhq) [inline db/migrate/20191209095548_ensures_unique_accepted_answer_post_id.rb]: For stable branch compatability, this should be
```suggestion
class EnsuresUniqueAcceptedAnswerPostId < ActiveRecord::Migration[5.2]
```

**Comment by SamSaffron** [comment]
I think I prefer this pattern: 

https://github.com/discourse/discourse/blob/33ae9679d8a05f415f430090313c990d2e532dbb/db/migrate/20151218232200_add_unique_index_to_category_users.rb#L6-L10

the correlated subquery is a bit harder to reason about.

**Comment by jjaffeux** [comment]
> I think I prefer this pattern:
> 
> https://github.com/discourse/discourse/blob/33ae9679d8a05f415f430090313c990d2e532dbb/db/migrate/20151218232200_add_unique_index_to_category_users.rb#L6-L10
> 
> the correlated subquery is a bit harder to reason about.
> 
> once you sort that feel free to merge.



> I think I prefer this pattern:
> 
> https://github.com/discourse/discourse/blob/33ae9679d8a05f415f430090313c990d2e532dbb/db/migrate/20151218232200_add_unique_index_to_category_users.rb#L6-L10
> 
> the correlated subquery is a bit harder to reason about.
> 
> once you sort that feel free to merge.

That's not the same thing though, this is to remove duplicate rows right? My query is to change one file with an array as value to one value.

**Comment by SamSaffron** [comment]
>  My query is to change one field with an array as value to one value as string.

I am not sure about that, I think it is doing the exact same. 

If we had to unique-ify values that are stored as `1,2,3` then we would need a different pattern altogether, custom fields that show up as arrays in our API are stored as multiple rows.

A good trick I tend to do when playing with this stuff is to have a "play" table on local, populate it with a bunch of stuff and then run the pattern by the table, it is very handy when dealing with the complex SQL tricks like the using statement.

**Comment by jjaffeux** [comment]
@SamSaffron ok updated PR.

Agreed it makes a lot more sense and is way more readable now that I have a better understanding of the data structure :+1:

Please review again as I don't do a lot of these lately.

File: db/migrate/20191209095548_ensures_unique_accepted_answer_post_id.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class EnsuresUniqueAcceptedAnswerPostId < ActiveRecord::Migration[5.2]
+  def change
+    execute <<~SQL
+      DELETE FROM topic_custom_fields AS tcf1
+      USING topic_custom_fields AS tcf2
+      WHERE tcf1.id > tcf2.id
+         AND tcf1.value = tcf2.value
```
Comment (SamSaffron) [inline db/migrate/20191209095548_ensures_unique_accepted_answer_post_id.rb]: this is getting close but not quite there ... I think the condition we want is: 

```
tcf1.id > tcf2.id AND
tcf1.topic_id = tcf2.topic_id AND
tcf1.name = tcf2.name AND
tcf1.name = 'accepted_answer_post_id'
```

We should test the query though by hand, I recommend that on local you have 2 topics, then accept answers on both, then jimmy the database by inserting a row, then run the query to clean up the mess and confirm both topics have a single row vs 2.

Comment (jjaffeux) [reply db/migrate/20191209095548_ensures_unique_accepted_answer_post_id.rb]: Yes I thought of it but wasn't sure if one was any better than the other.

This has been tested locally already

**Comment by jjaffeux** [comment]
My database tests:

### Before migration:

```
SELECT topic_id, count(*) FROM topic_custom_fields as tcf
WHERE tcf.name = 'accepted_answer_post_id'
GROUP BY tcf.topic_id
HAVING tcf.count > 1
```

| topic_id | count |
|----------|-------|
| 34236    | 2     |

```
SELECT count(*) FROM topic_custom_fields as tcf
WHERE tcf.name = 'accepted_answer_post_id'
```

| count |
|-------|
| 4230  |

```ruby
[1] pry(main)> Topic.find(34236).custom_fields['accepted_answer_post_id']
  Topic Load (4.1ms)  SELECT "topics".* FROM "topics" WHERE ("topics"."deleted_at" IS NULL) AND "topics"."id" = 34236 LIMIT 1
   (32.2ms)  SELECT "topic_custom_fields"."name", "topic_custom_fields"."value" FROM "topic_custom_fields" WHERE "topic_custom_fields"."topic_id" = 34236 ORDER BY id asc
=> ["146573", "146653"]
```

Migration:

```
== 20191209095548 EnsuresUniqueAcceptedAnswerPostId: migrating ================
-- execute("DELETE FROM topic_custom_fields AS tcf1\nUSING topic_custom_fields AS tcf2\nWHERE tcf1.id > tcf2.id AND\n   tcf1.topic_id = tcf2.topic_id AND\n   tcf1.name = tcf2.name AND\n   tcf1.name = 'accepted_answer_post_id'\n")
   -> 0.0888s
-- add_index(:topic_custom_fields, :topic_id, {:name=>:idx_topic_custom_fields_accepted_answer, :unique=>true, :where=>"name = 'accepted_answer_post_id'"})
   -> 0.0234s
== 20191209095548 EnsuresUniqueAcceptedAnswerPostId: migrated (0.1450s) =======
```

After migration:

```
SELECT topic_id, count(*) FROM topic_custom_fields as tcf
WHERE tcf.name = 'accepted_answer_post_id'
GROUP BY tcf.topic_id
HAVING tcf.count > 1
```

| topic_id | count |
|----------|-------|

```
SELECT count(*) FROM topic_custom_fields as tcf
WHERE tcf.name = 'accepted_answer_post_id'
```

| count |
|-------|
| 4229  |

```ruby
[2] pry(main)> Topic.find(34236).custom_fields['accepted_answer_post_id']
  Topic Load (6.6ms)  SELECT "topics".* FROM "topics" WHERE ("topics"."deleted_at" IS NULL) AND "topics"."id" = 34236 LIMIT 1
   (3.1ms)  SELECT "topic_custom_fields"."name", "topic_custom_fields"."value" FROM "topic_custom_fields" WHERE "topic_custom_fields"."topic_id" = 34236 ORDER BY id asc
=> "146573"
```

**Comment by SamSaffron** [comment]
That looks safe! Feel free to merge 🤗

**Pull request merged**